Sources/Fuzzilli/Corpus/MarkovCorpus.swift

```diff
@@ -0,0 +1,164 @@
+import Foundation
+
+/// Corpus & Scheduler based on 
+/// Coverage-based Greybox Fuzzing as Markov Chain paper
+/// https://mboehme.github.io/paper/TSE18.pdf
+/// Simply put, the corpus keeps track of which paths have been found, and prioritizes seeds
+/// whose path has been hit less than average. Ideally, this allows the fuzzer to prioritize
+/// less explored coverage
+
+public class MarkovCorpus: ComponentBase, Corpus {
+
+    private var allIncludedPrograms: [Program] // All programs currently in the corpus
+    private var programExecutionQueue: [Program]  // Queue of programs to be executed next, all of which hit a rare edge
+
+    private var edgeMap: [UInt64:Program]
+
+    private var totalExecs: UInt64
+
+    private var currentProg: Program?
+
+    private var remainingEnergy: UInt64
+
+    public init() {
+        self.allIncludedPrograms = []
+        self.programExecutionQueue = []
+        self.totalExecs = 0
+        self.edgeMap = [:]
+        self.currentProg = nil
+        self.remainingEnergy = 0
+        super.init(name: "MarkovCorpus")
+    }
+    
+    override func initialize() {
+        // The corpus must never be empty. Other components, such as the ProgramBuilder, rely on this
+        if isEmpty {
+            for _ in 1...5 {
+                let b = fuzzer.makeBuilder()
+                let objectConstructor = b.loadBuiltin("Object")
+                b.callFunction(objectConstructor, withArgs: [])
+                add(b.finalize())
+            }
+        }
+    }
+
+    /// Adds an individual program to the corpus.
+    /// This method should only be called for programs from outside sources, such as other connected workers, and imports from disk
+    /// to ensure that new edges are acquired/tracked properly.
+    public func add(_ program: Program) {
```

Comment:
Is this method (and `public func add(_ programs: [Program])`) currently used anywhere? I'm a bit worried that it will cause bugs later on because it depends on the program not having been passed to the evaluator ever before. I think I'd prefer if we just remove it entirely and make it clear that now programs can only be added to the corpus through the `public func add(_ program: Program, _ aspects: ProgramAspects)` API after they've been evaluated .

Reply:
I agree with your concerns. Taking a look at the codebase, the only outstanding usage of those two methods is in the Benchmarks file. I've modified that code to build a generic successful `ProgramAspects`, as well as the functionality internal to the corpus that builds an empty program. What do you think of that approach?

Reply:
I had to add similar code to ProgramBuilderTest as well.